File: Structural/Facade/Facade.py
from __future__ import annotations

from abc import ABC, abstractmethod


class IConverter(ABC):
    @abstractmethod
    def convert_video(self) -> None:
        pass


class Processor:
    def __init__(self, file):
        self._file = file

    def process_video(self):
        print(f"Processing video file {self._file}")


class Compressor:
    def __init__(self, file):
        self._file = file

    def compress_video(self):
        print(f"Compressing video file {self._file}")


class VideoConverter(IConverter):
    def __init__(self, file, video_format):
        self._file = file
        self._format = video_format

    def convert_video(self) -> str:
        Processor(self._file)
        Compressor(self._file)

        return f"Successfully converted video {self._file}"


video_converted = VideoConverter('video.mp4', "mpeg")
print(video_converted.convert_video())

File: Structural/Facade/Facade.go
package main

import "fmt"

type IConverter interface {
	convertVideo()
}

type VideoConverter struct {
	fileName string
}

func newConverter(fileName string) IConverter {
	return &VideoConverter{
		fileName: fileName,
	}
}

func (vc *VideoConverter) convertVideo() {
	newProcessor(vc.fileName).processVideo()
	newCompressor(vc.fileName).compressVideo()

	fmt.Println("Converting video file")
}

type Processor struct {
	fileName string
}

type IProcessor interface {
	processVideo()
}

// Process video
func newProcessor(fileName string) IProcessor {
	return &Processor{
		fileName: fileName,
	}
}

func (p *Processor) processVideo() {
	fmt.Println("Processing video file")
}

type Compressor struct {
	fileName string
}

type ICompressor interface {
	compressVideo()
}

// Compress video
func newCompressor(fileName string) ICompressor {
	return &Compressor{
		fileName: fileName,
	}
}

func (p *Compressor) compressVideo() {
	fmt.Println("Compressing video file")
}

func main() {
	videoConvert := newConverter("video.mp4")
	videoConvert.convertVideo()
}
